x-pack/plugins/rule_registry/server/plugin.ts

```diff
@@ -35,15 +66,43 @@ export class RuleRegistryPlugin implements Plugin<RuleRegistryPluginSetupContrac
       kibanaIndex: globalConfig.kibana.index,
       name: 'alerts',
       kibanaVersion: this.initContext.env.packageInfo.version,
-      logger: logger.get('root'),
+      logger: this.logger.get('root'),
       alertingPluginSetupContract: plugins.alerting,
-      writeEnabled: config.unsafe.write.enabled,
+      writeEnabled: this.config.unsafe.write.enabled,
     });
 
-    return rootRegistry;
+    return { rootRegistry };
   }
 
-  public start() {}
+  public start(
+    core: CoreStart,
+    plugins: RuleRegistryPluginStartDependencies
+  ): RuleRegistryPluginStartContract {
+    this.eventLogService = new EventLogService({
+      config: {
+        indexPrefix: this.config.indexPrefix,
+        isWriteEnabled: this.config.unsafe.write.enabled,
+      },
+      dependencies: {
+        clusterClient: Promise.resolve(core.elasticsearch.client), // TODO: get rid of Promise
+        spaces: plugins.spaces.spacesService,
```

Comment:
So that's a very legit question, and I have no ready answer at this point.

> Is there a hard dependency on the `RuleRegistryPluginStartDependencies` here?

Yes, `EventLogService` depends on `SpacesServiceStart` to be able to resolve space id by http request.

```ts
export interface SpacesServiceStart {
  /**
   * Retrieves the space id associated with the provided request.
   * @param request the request.
   */
  getSpaceId(request: KibanaRequest): string;
```

`SpacesServiceSetup` provides the same method, but it's deprecated and will be removed by 7.16

```ts
export interface SpacesServiceSetup {
  /**
   * Retrieves the space id associated with the provided request.
   * @param request the request.
   *
   * @deprecated Use `getSpaceId` from the `SpacesServiceStart` contract instead.
   * @removeBy 7.16
   */
  getSpaceId(request: KibanaRequest): string;
```

So I decided to use the start one and figure out how to fix the rest.

> If so, how do we inject the `eventLogService` into rule executors that are defined in the setup phase?

Not sure, but I have a few questions regarding this:

- How important is it to register rule types in the setup phase? Could we do it in the start phase? What would be differences / implications?
- In `security_solution` we depend on `SpacesServiceSetup` and we'll have to switch to `SpacesServiceStart` instead, so there's the same problem to solve there.
- Perhaps we could pass `SpacesServiceStart` dependency in the setup phase via `core.getStartServices()` - which returns a promise. That would require to change a few public methods of event log API from sync to async. I'd need to take a deeper look into that option.

Thanks for this question Marshall 👍  What do you think?

Reply:
Addressed. Moved to the `setup` phase, dependency on the spaces plugin is now a Promise, had to make `resolve` methods async.